taichi/codegen/codegen_opengl.cpp

```diff
@@ -153,10 +154,17 @@ class KernelGen : public IRVisitor {
     line_appender_header_.append_raw(kernel_header);
 
     int threads_per_group = opengl_get_threads_per_group();
-    if (num_threads_ < threads_per_group)
-      threads_per_group = std::max(1, num_threads_);
-    else
-      num_groups_ = (num_threads_ + threads_per_group - 1) / threads_per_group;
+    if (num_threads_ == -1) { // is dyn loop
+      num_groups_ = -1;
+    } else {
+      if (num_threads_ <= 0) num_threads_ = 1;
+      if (num_threads_ <= threads_per_group) {
+        threads_per_group = num_threads_;
```

Comment:
`threads_per_group` could be negative?

Reply:
It shouldn't be negative, unless OpenGL is broken:
https://github.com/taichi-dev/taichi/blob/cee79d9f043bafec4593d21b182b2727b3f26d45/taichi/backends/opengl/opengl_api.cpp#L450-L454

Reply:
In L160, `num_threads_ = -1`, and if the condition in L161 is true, then `threads_per_group = -1`.

Reply:
> In L160, num_threads_ = -1

You mean, `if (num_threads_ <= 0) num_threads_ = 1;`? It's `num_threads_ = 1`. Any negative num_threads_ will be eliminated in L160.

Reply:
Sorry.. For some reason I read L160 as `num_threads_ = -1`...